File: backend-java/src/main/java/com/example/calorie/body/dto/BodyDtos.java
package com.example.calorie.body.dto;

import com.example.calorie.body.BodyRecord;
import jakarta.validation.constraints.DecimalMax;
import jakarta.validation.constraints.DecimalMin;
import jakarta.validation.constraints.NotNull;

import java.math.BigDecimal;
import java.time.LocalDate;

public final class BodyDtos {

    private BodyDtos() {
    }

    public record BodyRecordRequest(
            @NotNull @DecimalMin("20") @DecimalMax("300") BigDecimal weightKg,
            @DecimalMin("1") @DecimalMax("70") BigDecimal bodyFatPct
    ) {
    }

    public record BodyRecordResponse(
            LocalDate recordedOn,
            BigDecimal weightKg,
            BigDecimal bodyFatPct
    ) {
        public static BodyRecordResponse from(BodyRecord record) {
            return new BodyRecordResponse(
                    record.getRecordedOn(), record.getWeightKg(), record.getBodyFatPct());
        }
    }
}

File: backend-kotlin/src/main/kotlin/com/example/calorie/body/dto/BodyDtos.kt
package com.example.calorie.body.dto

import com.example.calorie.body.BodyRecord
import jakarta.validation.constraints.DecimalMax
import jakarta.validation.constraints.DecimalMin
import java.math.BigDecimal
import java.time.LocalDate

data class BodyRecordRequest(
    @field:DecimalMin("20") @field:DecimalMax("300") val weightKg: BigDecimal,
    @field:DecimalMin("1") @field:DecimalMax("70") val bodyFatPct: BigDecimal? = null,
)

data class BodyRecordResponse(
    val recordedOn: LocalDate,
    val weightKg: BigDecimal,
    val bodyFatPct: BigDecimal?,
) {
    companion object {
        fun from(record: BodyRecord) =
            BodyRecordResponse(record.recordedOn, record.weightKg, record.bodyFatPct)
    }
}
